File: java/src/test/java/problems101_200/LinkedListCycleTest.java
package problems101_200;

import com.google.common.collect.Iterables;
import concepts.SinglyListNode;
import org.junit.Test;
import testing.Sample;
import testing.SampledTest;

import static com.google.common.truth.Truth.assertWithMessage;

public class LinkedListCycleTest extends SampledTest {
    @Test
    public void test() {
        for (LinkedListCycle solution : LinkedListCycle.values()) {
            for (TypedSample sample : getSamples(TypedSample[].class)) {
                SinglyListNode input = SinglyListNode.of(sample.input.head);
                if (sample.input.pos > -1) {
                    Iterables.getLast(input).next = Iterables.get(input, sample.input.pos);
                }
                assertWithMessage(sample.getMessage())
                    .that(solution.hasCycle(input))
                    .isEqualTo(sample.output);
            }
        }
    }

    static class TypedSample extends Sample<TypedSample.Input, Boolean> {
        static class Input {
            int[] head;
            int pos;
        }
    }
}

File: kotlin/src/test/kotlin/problems101_200/LinkedListCycleTest.kt
package problems101_200

import com.google.common.collect.Iterables
import com.google.common.truth.Truth.assertWithMessage
import concepts.SinglyListNode
import sample.SampledTest
import sample.component1
import sample.component2
import sample.component3
import sample.getSamples
import kotlin.test.Test

class LinkedListCycleTest : SampledTest() {
    @Test
    fun test() =
        LinkedListCycle.entries.forEach { solution ->
            getSamples<Array<LinkedListCycleSample>>().forEach { (input, output, message) ->
                assertWithMessage(message)
                    .that(
                        solution.hasCycle(
                            SinglyListNode.of(*input!!.head)!!.also {
                                if (input.pos > -1) {
                                    Iterables.getLast(it).next = Iterables.get(it, input.pos)
                                }
                            },
                        ),
                    ).isEqualTo(output)
            }
        }
}
